# 데이터 사전처리(Preprocessing)
- 머신러닝 등 데이터 분석의 정확도는 분석 데이터의 품질에 좌우된다.
- 데이터의 품질을 높이기 위해서 누락, 중복 데이터 등의 오류를 수정하고 분석 목적에 맞게 변형하는 과정이 필요하다.
- 수집한 데이터를 분석 목적에 적합하게 하는 작업을 사전처리(Preprocessing)이라고 한다.

### 누락 데이터(NaN : Not a Number)
- 데이터를 파일로 입력할 때나, 파일 형식을 변환하는 과정에서 데이터가 손실되는 경우가 발생
- 머신러닝 분석 모형에 데이터를 입력하기 전에 반드시 누락 데이터를 제거하고 다른 값으로 대체해야 한다.
- 누락 데이터가 많아지면 데이터의 품질이 떨어지고, 머신러닝 분석 알고리즘을 왜곡하는 현상이 발생한다.

In [302]:
import seaborn as sns
import pandas as pd

In [303]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [304]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


> deck에는 203개의 유효한 범주형 데이터만 있고 나머지 688개가 Nan

In [305]:
# deck열의 NaN 개수 파악하기, dropna : NaN 값은 제외
df['deck'].value_counts(dropna=False)

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

### 누락 데이터 찾기

In [306]:
# NaN 찾기
df.head().isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

In [307]:
# NaN이 아닌것 찾기
df.head().notnull()

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

In [308]:
# 누락 데이터 개수 확인
df.head().isnull().sum(axis=0)

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           3
embark_town    0
alive          0
alone          0
dtype: int64

In [309]:
# 누락 데이터 제거
missing_df = df.isnull()
missing_df

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [310]:
missing_df.deck.value_counts()

True     688
False    203
Name: deck, dtype: int64

In [311]:
for col in missing_df.columns:
    missing_count = missing_df[col].value_counts()
    try:
        print('%12s'%col, ':', missing_count[True])
    except:
        print('%12s'%col, ':', 0)
# missing_count

    survived : 0
      pclass : 0
         sex : 0
         age : 177
       sibsp : 0
       parch : 0
        fare : 0
    embarked : 2
       class : 0
         who : 0
  adult_male : 0
        deck : 688
 embark_town : 2
       alive : 0
       alone : 0


In [312]:
# NaN 값이 500개 이상인 열을 모두 삭제하겠다!
df_thresh = df.dropna(axis = 1, thresh=500)
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

> deck가 사라졌다!

In [313]:
# age 열에 나이 데이터가 없는 모든 행을 삭제
df_thresh_age = df_thresh.dropna(subset=['age'], how='any', axis=0)
len(df_thresh_age)

714

---
### 누락 데이터의 치환
- 누락 데이터를 무작정 삭제하면 수집한 데이터를 사용하지 못하게 될 수 있다.
- 데이터 중 일부가 누락되더라도, 나머지 데이터를 최대한 살려서 분석에 활용하는 경우가 많다.

In [314]:
df = sns.load_dataset('titanic')

In [315]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [316]:
# age 열의 10개 데이터 출력
df.age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [317]:
df[df.age.notnull()].age.mean()

29.69911764705882

In [318]:
# age 열의 NaN 값을 다른 나이 데이터의 평균으로 변경하기
df.age.fillna(round(df.age.mean()), inplace=True)

> mean()은 nan 값 빼고 계산한다.

In [319]:
df.age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    30.0
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [320]:
# embark_town의 829행 데이터 출력(NaN 위치)
df.embark_town[825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [321]:
# embark_town 열의 승선도시 중에서 가장 많이 출현한 값으로 바꾸기
most_freq = df['embark_town'].value_counts(dropna=False).idxmax() # idxmax : index의 최대값
most_freq

'Southampton'

In [322]:
df.embark_town.fillna(most_freq, inplace=True)

In [323]:
df.embark_town[825:830]
# df.embark_town.value_counts 로도 확인 가능

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object

### 앞뒤의 유사성 데이터로 값을 치환하기

In [324]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [325]:
df.embark_town[825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [326]:
# embark_town 열의 NaN 값을 바로 앞의 828 행의 값으로 바꾸기(ffill)
df.embark_town.fillna(method='ffill', inplace=True)

> 다음 값으로 바꾸려면 method='bfill'로 하면 된다.

In [327]:
df.embark_town[825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829     Queenstown
Name: embark_town, dtype: object

---
# 중복 데이터 처리
- 동일한 대상이 중복으로 존재하는 것은 분석 결과의 왜곡을 초래한다.

In [328]:
df = pd.DataFrame(
    {'c1':['a', 'a', 'b', 'a', 'b','a'],
    'c2':[1, 1, 1, 2, 2,1],
    'c3':[1, 1, 2, 2, 2,1]}
)
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2
5  a   1   1

In [329]:
# 데이터프레임 전체 행 데이터 중에서 중복값 찾기
df.duplicated()

0    False
1     True
2    False
3    False
4    False
5     True
dtype: bool

> df의 0행과 1행, 5행이 중복되므로 1행, 5행에 True로 출력됨

In [330]:
# 특정 열에서 데이터 중복값 찾기
df['c2'].duplicated()

0    False
1     True
2     True
3    False
4     True
5     True
Name: c2, dtype: bool

### 중복 데이터 제거

In [331]:
df.duplicated()
df.drop_duplicates()

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

> 1행과 5행이 삭제되었다.

In [332]:
# c2와 c3열을 기준으로 중복 행을 제거
df[['c2','c3']].duplicated()
df[['c2','c3']].drop_duplicates()

c2  c3
0   1   1
2   1   2
3   2   2

In [333]:
df[['c2','c3']].duplicated()
df.drop_duplicates(subset=['c2','c3'])

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

---
# 데이터 표준화
- 여러 곳에서 수집한 데이터는 서로 다른 단위가 섞이거나, 대상을 다른 형식으로 표현하는 경우가 많다.
- 대상을 표현하는 방법에 차이가 있으면 분석의 정확도는 현저히 낮아진다.
- 데이터 포맷을 일관성 있게 하는 작업이 표준화이다.

In [334]:
df = pd.read_csv('../data/auto-mpg.csv', header=None)
df

0  1      2      3       4     5   6  7                          8
0    18.0  8  307.0  130.0  3504.0  12.0  70  1  chevrolet chevelle malibu
1    15.0  8  350.0  165.0  3693.0  11.5  70  1          buick skylark 320
2    18.0  8  318.0  150.0  3436.0  11.0  70  1         plymouth satellite
3    16.0  8  304.0  150.0  3433.0  12.0  70  1              amc rebel sst
4    17.0  8  302.0  140.0  3449.0  10.5  70  1                ford torino
..    ... ..    ...    ...     ...   ...  .. ..                        ...
393  27.0  4  140.0  86.00  2790.0  15.6  82  1            ford mustang gl
394  44.0  4   97.0  52.00  2130.0  24.6  82  2                  vw pickup
395  32.0  4  135.0  84.00  2295.0  11.6  82  1              dodge rampage
396  28.0  4  120.0  79.00  2625.0  18.6  82  1                ford ranger
397  31.0  4  119.0  82.00  2720.0  19.4  82  1                 chevy s-10

[398 rows x 9 columns]

In [335]:
df.columns = ['mpg','cylinders','displacement','horsepower','weight','acceleration','model year','origin','name']
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0      130.0  3504.0          12.0   
1    15.0          8         350.0      165.0  3693.0          11.5   
2    18.0          8         318.0      150.0  3436.0          11.0   
3    16.0          8         304.0      150.0  3433.0          12.0   
4    17.0          8         302.0      140.0  3449.0          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

> mpg : mile per gallon ~ 한국과 단위가 다르다.

# 단위 변환
mpg(mile per gallon) >> kpl(kilometer per liter)
- 1 mile = 1.60934 km
- 1 gallon = 3.78541 liter

In [336]:
mpg_to_kpl = 1.60934 / 3.78541
df['kpl'] = round(df.mpg * mpg_to_kpl, 2)
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name   kpl  
0       1  chevrolet chevelle malibu  7.65  
1       1          buick skylark 320  6.38  
2       1         plymouth satellite  7.65  
3       1              amc rebel sst  6.80  
4       1                ford torino  7.23

### 자료형 변환

In [337]:
# 각 열의 자료형 확인
df.info()
df.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    float64
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
 9   kpl           398 non-null    float64
dtypes: float64(5), int64(3), object(2)
memory usage: 31.2+ KB


mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model year        int64
origin            int64
name             object
kpl             float64
dtype: object

> horsepower열이 object 자료형이다. 엔진 출력을 나타내는 열이므로 순자형으로 변환하는 것이 좋다.    
> 데이터를 보면 숫자처럼 보이지만 자료형이 다른 것으로 나오는 것을 보면 데이터 중에 다른 자료형이 있음을 알 수 있다.

In [338]:
# horsepower 열의 고유값 확인
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [339]:
# data 개수 확인
len(df.horsepower)

398

In [340]:
# data가 ?인 것을 확인
mask = df['horsepower'] == '?'
mask

0      False
1      False
2      False
3      False
4      False
       ...  
393    False
394    False
395    False
396    False
397    False
Name: horsepower, Length: 398, dtype: bool

In [341]:
df[mask]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model year  origin                  name    kpl  
32           71       1            ford pinto  10.63  
126          74       1         ford maverick   8.93  
330          80       2  renault lecar deluxe  17.39  
336          80       1    ford mustang cobra  10.03  
354          81       2           renault 18i  14.67  
374          82       1        amc concord dl   9.78

In [342]:
# 누락 데이터(?)를 삭제 : ?를 np.nan으로 변경하고 dropna
import numpy as np

In [343]:
df.horsepower.replace('?', np.nan, inplace=True)

In [344]:
df.horsepower.unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', nan, '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

> nan 값이 생겼다.

In [345]:
# 누락 데이터 삭제
df.dropna(subset=['horsepower'], axis=0, inplace=True)

In [346]:
len(df['horsepower'])

392

In [347]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    object 
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(5), int64(3), object(2)
memory usage: 33.7+ KB


In [348]:
# horsepower type을 실수형으로 변경
df['horsepower'] = df.horsepower.astype('float')

In [349]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(6), int64(3), object(1)
memory usage: 33.7+ KB


In [350]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year  origin                       name   kpl  
0          70       1  chevrolet chevelle malibu  7.65  
1          70       1          buick skylark 320  6.38  
2          70       1         plymouth satellite  7.65  
3          70       1              amc rebel sst  6.80  
4          70       1                ford torino  7.23

In [351]:
df.origin.value_counts()

1    245
3     79
2     68
Name: origin, dtype: int64

In [352]:
# origin의 데이터를 문자형 데이터로 변형(1:USA, 2:EU, 3:JAPAN)
df.origin.replace([1,2,3], ['USA','EU','JAPAN']).value_counts()

USA      245
JAPAN     79
EU        68
Name: origin, dtype: int64

In [353]:
# df.origin.replace([1,2,3], ['USA','EU','JAPAN'], inplace=True)

In [354]:
df.origin.replace({1:'USA', 2:'EU', 3:'JAPAN'}, inplace=True)

> dictionary 도 가능하다

In [355]:
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year origin                       name   kpl  
0          70    USA  chevrolet chevelle malibu  7.65  
1          70    USA          buick skylark 320  6.38  
2          70    USA         plymouth satellite  7.65  
3          70    USA              amc rebel sst  6.80  
4          70    USA                ford torino  7.23

In [356]:
df.origin.describe()

count     392
unique      3
top       USA
freq      245
Name: origin, dtype: object

In [357]:
# origin 열의 문자열 자료형을 범주형으로 변환
df['origin'] = df.origin.astype('category')
df.origin.dtypes

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False)

In [358]:
# 범주형을 문자형으로 변환
df.origin = df.origin.astype('str')
df.origin.dtypes

dtype('O')

In [362]:
# model year가 정수로 되어있으나 수치계산과 관련 없으므로 범주형으로 변환
df['model year'].dtypes

dtype('int64')

In [363]:
df['model year'] = df['model year'].astype('category')
df['model year'].dtypes

CategoricalDtype(categories=[70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82], ordered=False)

---
### 범주형(카테고리) 데이터 처리
- 연속 데이터를 그대로 사용하기보다는 일정한 구간(binning)으로 나눠서 분석하는 것이 효율적인 경우가 많다.

In [372]:
# horsepower를 3개의 범위로 나누기
count, bin_dividers = np.histogram(df['horsepower'], bins=3)
count, bin_dividers

(array([257, 103,  32]),
 array([ 46.        , 107.33333333, 168.66666667, 230.        ]))

In [373]:
bin_dividers

array([ 46.        , 107.33333333, 168.66666667, 230.        ])

> 257, 103, 32 부분은 count값, 46~107/107~168/168~230 부분은 범위를 나눠준 것.

In [365]:
# 3개의 bin에 이름 지정
bin_names = ['저출력','보통출력','고출력']

In [368]:
# cut by bin
df['hp_bin'] = pd.cut(
    x = df['horsepower'],
    bins=bin_dividers,
    labels=bin_names,
    include_lowest=True # 첫 경계값 포함
)

In [371]:
df[['horsepower', 'hp_bin']].head(15)

horsepower hp_bin
0        130.0   보통출력
1        165.0   보통출력
2        150.0   보통출력
3        150.0   보통출력
4        140.0   보통출력
5        198.0    고출력
6        220.0    고출력
7        215.0    고출력
8        225.0    고출력
9        190.0    고출력
10       170.0    고출력
11       160.0   보통출력
12       150.0   보통출력
13       225.0    고출력
14        95.0    저출력

> 머신러닝에서 범주형의 문자열을 쓸 수 없다!

---
### 더미(dummy) 변수
- Batch : 일괄처리 (그냥 참고)
- 범주형 데이터를 회귀분석이나 머신러닝에 바로 사용할 수 없는 경우
- 0 또는 1로 표현
- one-hot encoding과 유사하다

In [374]:
horsepower_dummies = pd.get_dummies(df['hp_bin'])
horsepower_dummies

저출력  보통출력  고출력
0      0     1    0
1      0     1    0
2      0     1    0
3      0     1    0
4      0     1    0
..   ...   ...  ...
393    1     0    0
394    1     0    0
395    1     0    0
396    1     0    0
397    1     0    0

[392 rows x 3 columns]

---
### One-Hot Encoding
- 범주형 데이터를 0,1 을 갖는 원핫벡터로 변환
- 결과는 선형대수학에서 정의하는 희소행렬(sparse matrix)로 변환
- 희소행렬은 행/열 좌표와 값 형태로 정리

In [375]:
from sklearn import preprocessing # 사이키런

In [377]:
# 전처리를 위한 encoder 객체 만들기
label_encoder = preprocessing.LabelEncoder() # label encoder 생성
onehot_encoder = preprocessing.OneHotEncoder()

In [385]:
df['hp_bin'].head(15)

0     보통출력
1     보통출력
2     보통출력
3     보통출력
4     보통출력
5      고출력
6      고출력
7      고출력
8      고출력
9      고출력
10     고출력
11    보통출력
12    보통출력
13     고출력
14     저출력
Name: hp_bin, dtype: category
Categories (3, object): ['저출력' < '보통출력' < '고출력']

In [386]:
# label encoder로 문자열 범주를 숫자형 범주로 바꾸기
onehot_labeled=label_encoder.fit_transform(df['hp_bin'].head(15))
onehot_labeled

array([1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 2])

> 보통출력 : 1, 고출력 : 0, 저출력 : 2

In [382]:
# 2차원 행렬로 형태를 변경하자 ~ 희소행렬의 좌표값으로 쓰기위해
onehot_reshaped = onehot_labeled.reshape(len(onehot_labeled),1)
onehot_reshaped

array([[1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [2]])

In [384]:
# 희소행렬
onehot_fitted = onehot_encoder.fit_transform(onehot_reshaped)
print(onehot_fitted)

  (0, 1)	1.0
  (1, 1)	1.0
  (2, 1)	1.0
  (3, 1)	1.0
  (4, 1)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 0)	1.0
  (11, 1)	1.0
  (12, 1)	1.0
  (13, 0)	1.0
  (14, 2)	1.0


> 각 좌표에 1 값이 들어있다.    
> 각 좌표의 열 값이 onehot_labeled에 저장된 값으로 되어있다.

---
### 정규화(Normalization)
- 각 열에 있는 숫자 데이터의 상대적 크기 때문에, 머신러닝 분석결과가 달라질 수 있다.
- 숫자 데이터의 상대적인 크기 차이를 제거

In [387]:
# 최대값의 절대값을 이용한 정규화
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [389]:
# horsepower 열의 최대값의 절대값(abs)으로 모든 데이터를 나누기 : 0과 1 사이의 값으로 변환
horsepower1 = df.horsepower / abs(df.horsepower.max())
horsepower1.describe()

count    392.000000
mean       0.454215
std        0.167353
min        0.200000
25%        0.326087
50%        0.406522
75%        0.547826
max        1.000000
Name: horsepower, dtype: float64

In [390]:
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [391]:
# 각 열의 데이터에서 최소값을 뺀 값을 분자로
min_x = df.horsepower - df.horsepower.min()

# 각 열의 최대값에서 최소값을 뺀 값을 분모로
min_max = df.horsepower.max() - df.horsepower.min()

# min_x / min_max 는 0과 1 사이의 값이 나오게 된다.
horsepower2 = min_x / min_max
horsepower2.describe()

count    392.000000
mean       0.317768
std        0.209191
min        0.000000
25%        0.157609
50%        0.258152
75%        0.434783
max        1.000000
Name: horsepower, dtype: float64

---
# 시계열 데이터
- 특정한 시점을 기록하는 Timestamp
- 두 지점 사이의 일정한 기간ㅇ르 나타내는 Period

In [392]:
df = pd.read_csv('../data/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [393]:
df.isnull().sum()

Date      0
Close     0
Start     0
High      0
Low       0
Volume    0
dtype: int64

In [394]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [398]:
df['new_Date'] = pd.to_datetime(df['Date'])
df.head()

Date  Close  Start   High    Low  Volume   new_Date
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26

In [399]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Date      20 non-null     object        
 1   Close     20 non-null     int64         
 2   Start     20 non-null     int64         
 3   High      20 non-null     int64         
 4   Low       20 non-null     int64         
 5   Volume    20 non-null     int64         
 6   new_Date  20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.2+ KB


In [404]:
# new_Date 인덱스로 만들기
df.set_index('new_Date', inplace=True)

In [410]:
# 기존 Date열 삭제
# df.iloc[:,1:]
df.drop('Date', axis=1, inplace=True)

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [411]:
df.head()

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039

In [412]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 20 entries, 2018-07-02 to 2018-06-01
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Close   20 non-null     int64
 1   Start   20 non-null     int64
 2   High    20 non-null     int64
 3   Low     20 non-null     int64
 4   Volume  20 non-null     int64
dtypes: int64(5)
memory usage: 960.0 bytes


### Timestamp를 Period로 변환

In [413]:
# 날짜 형식의 문자열로 구성되는 리스트 정의
dates = ['2019-01-01', '2020-03-01', '2021-06-01']

In [414]:
ts_dates = pd.to_datetime(dates)
ts_dates

DatetimeIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='datetime64[ns]', freq=None)

In [417]:
pr_day = ts_dates.to_period(freq='D')
print(pr_day)
pr_month = ts_dates.to_period(freq='M')
print(pr_month)
pr_year = ts_dates.to_period(freq='Y') # Y 혹은 A
print(pr_year)

PeriodIndex(['2019-01-01', '2020-03-01', '2021-06-01'], dtype='period[D]')
PeriodIndex(['2019-01', '2020-03', '2021-06'], dtype='period[M]')
PeriodIndex(['2019', '2020', '2021'], dtype='period[A-DEC]')


> freq Option   
> D (Day: 1일)   
> W (Week:1주)   
> M (Month end: 월말)   
> MS (Month begin: 월초)   
> Q (Quarter end : 분기말)   
> QS (Quarter begin: 분기초)   
> A (year end : 연말)   
> AS (year begin: 연초)   
> B (Business day : 휴일 제외)   
> H (hour : 1시간)   
> T (minute :1분)   
> S (second: 1초)

### 시계열 데이터 만들기

In [418]:
# Timestamp의 배열 만들기 : 월 간격, 월 시작일 기준
ts_ms = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날짜 범위의 끝(미정)
    periods=6, # 생성할 Timestamp의 개수
    freq='MS', # 시간 간격 : 월초
    tz='Asia/Seoul', # timezone 시간대
)
ts_ms

DatetimeIndex(['2022-01-01 00:00:00+09:00', '2022-02-01 00:00:00+09:00',
               '2022-03-01 00:00:00+09:00', '2022-04-01 00:00:00+09:00',
               '2022-05-01 00:00:00+09:00', '2022-06-01 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='MS')

In [419]:
# 월 간격, 월 마지막날 기준
ts_me = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날짜 범위의 끝(미정)
    periods=6, # 생성할 Timestamp의 개수
    freq='M', # 시간 간격 : 월말
    tz='Asia/Seoul', # timezone 시간대
)
ts_me

DatetimeIndex(['2022-01-31 00:00:00+09:00', '2022-02-28 00:00:00+09:00',
               '2022-03-31 00:00:00+09:00', '2022-04-30 00:00:00+09:00',
               '2022-05-31 00:00:00+09:00', '2022-06-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='M')

In [422]:
# 분기(3개월) 간격, 월의 마지막날 기준
ts_3m = pd.date_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날짜 범위의 끝(미정)
    periods=6, # 생성할 Timestamp의 개수
    freq='3M', # 시간 간격 : 3개월 말
    tz='Asia/Seoul', # timezone 시간대
)
ts_3m

DatetimeIndex(['2022-01-31 00:00:00+09:00', '2022-04-30 00:00:00+09:00',
               '2022-07-31 00:00:00+09:00', '2022-10-31 00:00:00+09:00',
               '2023-01-31 00:00:00+09:00', '2023-04-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='3M')

---
### Period 배열


In [423]:
# 1개월 길이
pr_m = pd.period_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날씨 범위의 끝
    periods=3, # 생성할 period 수
    freq='M'
)
pr_m

PeriodIndex(['2022-01', '2022-02', '2022-03'], dtype='period[M]')

In [424]:
# 1시간 길이
pr_h = pd.period_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날씨 범위의 끝
    periods=3, # 생성할 period 수
    freq='H'
)
pr_h

PeriodIndex(['2022-01-01 00:00', '2022-01-01 01:00', '2022-01-01 02:00'], dtype='period[H]')

In [425]:
# 2시간 길이
pr_h = pd.period_range(
    start='2022-01-01', # 날짜 범위의 시작
    end=None, # 날씨 범위의 끝
    periods=3, # 생성할 period 수
    freq='2H' # 2시간 간격
)
pr_h

PeriodIndex(['2022-01-01 00:00', '2022-01-01 02:00', '2022-01-01 04:00'], dtype='period[2H]')

In [427]:
df

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039
2018-06-25  11150  11400  11450  11000   55519
2018-06-22  11300  11250  11450  10750  134805
2018-06-21  11200  11350  11750  11200  133002
2018-06-20  11550  11200  11600  10900  308596
2018-06-19  11300  11850  11950  11300  180656
2018-06-18  12000  13400  13400  12000  309787
2018-06-15  13400  13600  13600  12900  201376
2018-06-14  13450  13200  13700  13150  347451
2018-06-12  13200  12200  13300  12050  558148
2018-06-11  11950  12000  12250  11950   62293
2018-06-08  11950  11950  12200  11800   59258
2018-06-07  11950  12200  12300  11900   49088
2018-06-05  12150  11800  12250  11800   42485
2018-06-04  11900  11900  12200  11700   25171
2018-06-01  11900  11800  12100  11750   32062

In [429]:
# new_Date 인덱스 해제
df.reset_index(inplace=True)

In [430]:
df.head()

new_Date  Close  Start   High    Low  Volume
0 2018-07-02  10100  10850  10900  10000  137977
1 2018-06-29  10700  10550  10900   9990  170253
2 2018-06-28  10400  10900  10950  10150  155769
3 2018-06-27  10900  10800  11050  10500  133548
4 2018-06-26  10800  10900  11000  10700   63039

In [432]:
# dt.year : data type year
df['Year'] = df['new_Date'].dt.year
df['Month'] = df['new_Date'].dt.month
df['Day'] = df['new_Date'].dt.day
df.head()

new_Date  Close  Start   High    Low  Volume  Year  Month  Day
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26

In [434]:
df['Date_yr'] = df.new_Date.dt.to_period(freq='A') # A : year
df['Date_m'] = df.new_Date.dt.to_period(freq='M') # M : year + month
df.head()

new_Date  Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
0 2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
1 2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2 2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
3 2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
4 2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   

    Date_m  
0  2018-07  
1  2018-06  
2  2018-06  
3  2018-06  
4  2018-06

In [435]:
df.set_index('new_Date', inplace=True)

In [436]:
df.head()

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   

             Date_m  
new_Date             
2018-07-02  2018-07  
2018-06-29  2018-06  
2018-06-28  2018-06  
2018-06-27  2018-06  
2018-06-26  2018-06

In [437]:
# 날짜 인덱스를 이용하여 데이터 선택하기
df_y = df.loc['2018']
df_y

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   
2018-06-29  10700  10550  10900   9990  170253  2018      6   29    2018   
2018-06-28  10400  10900  10950  10150  155769  2018      6   28    2018   
2018-06-27  10900  10800  11050  10500  133548  2018      6   27    2018   
2018-06-26  10800  10900  11000  10700   63039  2018      6   26    2018   
2018-06-25  11150  11400  11450  11000   55519  2018      6   25    2018   
2018-06-22  11300  11250  11450  10750  134805  2018      6   22    2018   
2018-06-21  11200  11350  11750  11200  133002  2018      6   21    2018   
2018-06-20  11550  11200  11600  10900  308596  2018      6   20    2018   
2018-06-19  11300  11850  11950  11300  180656  2018      6   19    2018   
2018-06-18  12000  13400  13400  12000  309787  2018      6   18    2018   
2018-06-15  13400  13600  13600  12900  201376  2018      6   15    2018   
2018-06-14  13450  13200  13700  13150  347451  2018      6   14    2018   
2018-06-12  13200  12200  13300  12050  558148  2018      6   12    2018   
2018-06-11  11950  12000  12250  11950   62293  2018      6   11    2018   
2018-06-08  11950  11950  12200  11800   59258  2018      6    8    2018   
2018-06-07  11950  12200  12300  11900   49088  2018      6    7    2018   
2018-06-05  12150  11800  12250  11800   42485  2018      6    5    2018   
2018-06-04  11900  11900  12200  11700   25171  2018      6    4    2018   
2018-06-01  11900  11800  12100  11750   32062  2018      6    1    2018   

             Date_m  
new_Date             
2018-07-02  2018-07  
2018-06-29  2018-06  
2018-06-28  2018-06  
2018-06-27  2018-06  
2018-06-26  2018-06  
2018-06-25  2018-06  
2018-06-22  2018-06  
2018-06-21  2018-06  
2018-06-20  2018-06  
2018-06-19  2018-06  
2018-06-18  2018-06  
2018-06-15  2018-06  
2018-06-14  2018-06  
2018-06-12  2018-06  
2018-06-11  2018-06  
2018-06-08  2018-06  
2018-06-07  2018-06  
2018-06-05  2018-06  
2018-06-04  2018-06  
2018-06-01  2018-06

In [438]:
# 날짜 인덱스를 이용하여 데이터 선택하기
df_y = df.loc['2018-07']
df_y

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
new_Date                                                                   
2018-07-02  10100  10850  10900  10000  137977  2018      7    2    2018   

             Date_m  
new_Date             
2018-07-02  2018-07

In [439]:
# 날짜 인덱스를 이용하여 데이터 선택하기
df_y = df.loc['2018-07', 'Start':'High']
df_y

Start   High
new_Date                
2018-07-02  10850  10900

In [440]:
# 시간 간격 계산 : 최근 180일 ~ 189일 사이의 값들 선택하기
today = pd.to_datetime('2018-12-25')
df['time_delta'] = today - df.index # index 가 datatime 형태, : 결과가 ~ days 형태
df.set_index('time_delta', inplace=True)

In [442]:
df_180 = df.loc['180 days':'189 days']
df_180

Close  Start   High    Low  Volume  Year  Month  Day Date_yr  \
time_delta                                                                 
180 days    10400  10900  10950  10150  155769  2018      6   28    2018   
181 days    10900  10800  11050  10500  133548  2018      6   27    2018   
182 days    10800  10900  11000  10700   63039  2018      6   26    2018   
183 days    11150  11400  11450  11000   55519  2018      6   25    2018   
186 days    11300  11250  11450  10750  134805  2018      6   22    2018   
187 days    11200  11350  11750  11200  133002  2018      6   21    2018   
188 days    11550  11200  11600  10900  308596  2018      6   20    2018   
189 days    11300  11850  11950  11300  180656  2018      6   19    2018   

             Date_m  
time_delta           
180 days    2018-06  
181 days    2018-06  
182 days    2018-06  
183 days    2018-06  
186 days    2018-06  
187 days    2018-06  
188 days    2018-06  
189 days    2018-06